src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,295 @@
+
+      if (typed_value_segment != nullptr) {
+        // For ValueSegments simply compare each value of the value segment
+        // against the compare value with given comparator
+        std::vector<Type> values = typed_value_segment->values();
+        auto comparator = _get_comparator<Type>(_scan_type);
+
+        for (auto cell_id = 0ul, typed_segment_size = values.size(); cell_id < typed_segment_size; ++cell_id) {
+          if (comparator(values[cell_id], typed_search_value)) {
+            position_list.push_back(RowID{static_cast<ChunkID>(chunk_id), static_cast<ChunkOffset>(cell_id)});
+          }
+        }
+      } else if (const auto typed_dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+                 typed_dictionary_segment != nullptr) {
+        /**
+         * For Dictionary Segments use index-based comparing with value ids.
+         * This is possible because dictionary values are stored in sorted manner and the total order of dictionary value ids 
+         * therefore is based on the total order of values in the attribute vector.
+         * Specifically, value_id_a <= value_id_b iff. value_a <= value_b.
+         * We compare by first retrieving the correct compare_value_id to compare against given the ScanType and search value, 
+         * and then comparing the ValueIDs in the attribute vector against it.
+         */
+
+        auto dictionary_comparator = _get_comparator<ValueID>(_scan_type);
+        auto attribute_value_ids = typed_dictionary_segment->attribute_vector();
+
+        const auto search_value_id =
+            _get_compare_value_id(_scan_type, typed_dictionary_segment->upper_bound(typed_search_value),
+                                  typed_dictionary_segment->lower_bound(typed_search_value));
+
+        if (search_value_id == EMPTY_RESULT) {
+          continue;
+        }
+
+        if (search_value_id == SELECT_EVERYTHING) {
+          for (auto cell_id = 0ul, typed_segment_size = attribute_value_ids->size(); cell_id < typed_segment_size;
+               ++cell_id) {
+            position_list.push_back(RowID{static_cast<ChunkID>(chunk_id), static_cast<ChunkOffset>(cell_id)});
+          }
+          continue;
+        }
+
+        // Compare each value_id in the attribute vector to the calculated compare_value_id with comparator
+        for (auto cell_id = 0ul, typed_segment_size = attribute_value_ids->size(); cell_id < typed_segment_size;
+             ++cell_id) {
+          if (dictionary_comparator(attribute_value_ids->get(cell_id), search_value_id)) {
+            position_list.push_back(RowID{static_cast<ChunkID>(chunk_id), static_cast<ChunkOffset>(cell_id)});
+          }
+        }
+      } else if (const auto typed_reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+                 typed_reference_segment != nullptr) {
+        // For ReferenceSegments we materialize each referenced value and compare it to the search value
```

Comment:
As explained in the lecture, the idea was to retrieve the underlying Value- or DictionarySegment and avoid the costly `operator[]`.